draco/asp/hard.lp

```diff
@@ -39,8 +117,35 @@ violation(count_with_field) :-
     attribute((encoding,aggregate),E,count),
     attribute((encoding,field),E,_).
 
+% @hard(text_mark_without_text_channel) Text mark requires text encoding.
+violation(text_mark_without_text_channel) :-
+    attribute((mark,type),M,text),
+    not attribute(mark_channel,M,text).
+
+% @hard(text_channel_without_text_mark) Text channel requires text mark.
+violation(text_channel_without_text_mark) :-
+    attribute(mark_channel,M,text),
+    not attribute((mark,type),M,text).
+
 % @hard(point_tick_bar_without_x_or_y) Point, tick, and bar require x or y channel.
 violation(point_tick_bar_without_x_or_y) :-
     attribute((mark,type),M,(point;tick;bar)),
     not attribute(mark_channel,M,x),
     not attribute(mark_channel,M,y).
+
+% @hard(line_area_without_x_y) Line and area require x and y channel.
+violation(line_area_without_x_y) :-
+    attribute((mark,type),M,(line;area)),
+    not attribute(mark_x_y_channels,M).
+
+% @hard(line_area_with_discrete) Line and area cannot have two discrete(ie. x and y cannot both be discrete).
+violation(line_area_with_discrete) :-
+    attribute((mark,type),M,(line;area)),
+    attribute(discrete,M,x),
+    attribute(discrete,M,y).
+
+% @hard(bar_tick_continuous_x_y) Bar and tick cannot have both x and y continuous.
+violation(bar_tick_continuous_x_y) :-
+    attribute((mark,type),M,(tick;bar)),
+    attribute(mark_cont_channel,M,x),
```

Comment:
This helper name is very different from `discrete` above, although it is similar.

Reply:
I wonder whether it's cleaner to write constraints over scale types. Write a helper to add the scale type to an encoding and then ensure that the scale type is continuous. For the later, you can have a fact like `domain(continuous,(linear;log))`.